nixos/modules/security/pam/default.nix

```
@@ -0,0 +1,190 @@
+# This module provides configuration for the PAM (Pluggable
+# Authentication Modules) system.
+
+{ config, lib, pkgs, ... }:
+
+with lib;
+
+let
+  cfg = config.security.pam;
+  pamLib = cfg.lib;
+
+  pamEntryModule = entryType: { config, ... }: {
+    options = {
+      entryType = mkOption {
+        type = types.enum [ "account" "auth" "password" "session" ];
+        default = entryType;
+      };
+
+      control = mkOption {
+        type = pamLib.controlType;
+        description = ''
+          The PAM control for this module. It can be either on of the
+          historical basic control keywords, of a set of result-actions pairs.
+        '';
+        apply = value: if isString value then value else
+                      ("[" + concatStringsSep " " (mapAttrsToList (n: v: "${n}=${toString v}") value) + "]");
+      };
+
+      path = mkOption {
+        type = types.str;
+        description = "The path to this module shared library";
+      };
+
+      args = mkOption {
+        default = [];
+        type = types.listOf types.str;
+        description = "Arbitrary string arguments passed to this modules.";
+        apply = value: concatStringsSep " " (map toString value);
```

As commented on [here](https://github.com/NixOS/nixpkgs/pull/105098#discussion_r532118632), I think we should be escaping arguments that contain spaces properly here.

---

```suggestion
        apply = value: concatStringsSep " " (map (v: if hasInfix " " v then "[${escape [ "]" ] v}]" else v;) value);
```

Does something like this look okay to you?

---

Yeah, LGTM.